lib/discourse_redis.rb

```diff
@@ -150,6 +150,26 @@ def self.new_redis_store
     Cache.new
   end
 
+  def multi
+    if block_given?
+      @redis.multi do |transaction|
+        yield DiscourseRedis.new(@config, namespace: @namespace, raw_redis: transaction)
+      end
+    else
+      @redis.multi
+    end
+  end
+
+  def pipelined
+    if block_given?
+      @redis.pipelined do |transaction|
+        yield DiscourseRedis.new(@config, namespace: @namespace, raw_redis: transaction)
+      end
+    else
+      @redis.pipelined
+    end
+  end
```

Comment:
```suggestion
  def pipelined(&block)
    if block.arity == 0 
      Discourse.deprecate(<<~MSG, output_in_test: true, drop_from: '3.0')
      Accessing the global Redis instance within `DiscourseRedis#pipelined` is deprecated.

      ```
      Discourse.Redis.pipelined do 
        Discourse.Redis.get("somekey")
      end
      ```

      replaced by 

      ```
      Discourse.Redis.pipelined do |transaction|
        transaction.get("somekey")
      end
      MSG
    end

    @redis.pipelined do |transaction|
      yield DiscourseRedis.new(@config, namespace: @namespace, raw_redis: transaction)
    end
  end
```

I don't think we should even be supporting using `DiscourseRedis#pipelined` without a block. Also, I added a deprecation warning to log other instances where we might still be accessing the global Redis client within the pipelined block. This is similar to what redis-rb is doing to warn people about the change in usage.

Reply:
I guess we would need this for multi as well? If we do this though are we overdoing it since the underlying client already handles the deprecation warnings?

Reply:
Ahhh I had a look at `redis-rb`'s implementation and it does log a deprecation warning if you try to access another Redis client outside of the `pipelined` block.